controllers/volumereplicationgroup_controller.go

```diff
@@ -1461,6 +1507,73 @@ func (v *VRGInstance) createOrUpdateVR(vrNamespacedName types.NamespacedName,
 	return v.updateVR(volRep, state, log)
 }
 
+func (v *VRGInstance) updateMetricsTimer(desiredState ramendrv1alpha1.State) {
+	if desiredState == "" {
+		return
+	}
+
+	if val, ok := metricsTimersMap[v.instance.Name]; !ok {
+		metricsTimersMap[v.instance.Name] = timerInstance{running: false, state: desiredState} // initialize instance
+	} else {
+		if val.state != desiredState { // start timer
+			v.startTimer(desiredState)
+		} else { // stop timer when status finished updating
+			v.stopTimer(desiredState)
+		}
+	}
+}
+
+func (v *VRGInstance) startTimer(desiredState ramendrv1alpha1.State) {
+	switch desiredState {
+	case ramendrv1alpha1.PrimaryState:
+		v.metricsTimerStart(metricsTimeToPrimary, desiredState)
+	case ramendrv1alpha1.SecondaryState:
+		v.metricsTimerStart(metricsTimeToSecondary, desiredState)
+	case ramendrv1alpha1.UnknownState:
+		fallthrough
+	default:
+		// not supported
+	}
+}
+
+func (v *VRGInstance) stopTimer(desiredState ramendrv1alpha1.State) {
+	switch desiredState {
+	case ramendrv1alpha1.PrimaryState:
+		fallthrough
+	case ramendrv1alpha1.SecondaryState:
+		v.metricsTimerStop()
+	case ramendrv1alpha1.UnknownState:
+		fallthrough
+	default:
+		// not supported
+	}
+}
+
+func (v *VRGInstance) metricsTimerStart(histogram prometheus.Histogram, desiredVRGState ramendrv1alpha1.State) {
+	if val, ok := metricsTimersMap[v.instance.Name]; ok {
+		if !val.running {
+			val.state = desiredVRGState
+			val.timer.ObserveDuration() // stop existing timer if one is running
```

Comment:
This was updated to be a bit cleaner, but I'll elaborate here. `startTimer` and `metricsTimerStop` have to deal with spurious reconciles on occasion (though this may have been fixed by now), so if there's a timer running already and `startTimer` is called, just ignore it and continue. Likewise for `metricsTimerStop`: the timer has to be non-nil and the desiredState must be different than the current state.